src/crimson/osd/osd_operations/client_request.cc

```diff
@@ -230,74 +230,73 @@ ClientRequest::process_op(
+  );
+  co_await recover_missings(pg, m->get_hobj(), snaps_need_to_recover());
+
+  DEBUGDPP("{}.{}: checking already_complete",
+	   *pg, *this, this_instance_id);
+  auto completed = co_await pg->already_complete(m->get_reqid());
+
+  if (completed) {
+    DEBUGDPP("{}.{}: already completed, sending reply",
 	     *pg, *this, this_instance_id);
-    return pg->already_complete(m->get_reqid()).then_interruptible(
-      [FNAME, this, pg, this_instance_id, &ihref](auto completed) mutable
-      -> PG::load_obc_iertr::future<> {
-      if (completed) {
-	DEBUGDPP("{}.{}: already completed, sending reply",
-		 *pg, *this, this_instance_id);
-        auto reply = crimson::make_message<MOSDOpReply>(
-          m.get(), completed->err, pg->get_osdmap_epoch(),
-          CEPH_OSD_FLAG_ACK | CEPH_OSD_FLAG_ONDISK, false);
-	reply->set_reply_versions(completed->version, completed->user_version);
-        // TODO: gate the crosscore sending
-        return get_foreign_connection().send_with_throttling(std::move(reply));
-      } else {
-	DEBUGDPP("{}.{}: not completed, entering get_obc stage",
-		 *pg, *this, this_instance_id);
-        return ihref.enter_stage<interruptor>(client_pp(*pg).get_obc, *this
-	).then_interruptible(
-          [FNAME, this, pg, this_instance_id, &ihref]() mutable
-	  -> PG::load_obc_iertr::future<> {
-	  DEBUGDPP("{}.{}: entered get_obc stage, about to wait_scrub",
+    auto reply = crimson::make_message<MOSDOpReply>(
+      m.get(), completed->err, pg->get_osdmap_epoch(),
+      CEPH_OSD_FLAG_ACK | CEPH_OSD_FLAG_ONDISK, false);
+    reply->set_reply_versions(completed->version, completed->user_version);
+    // TODO: gate the crosscore sending
+    co_await interruptor::make_interruptible(
+      get_foreign_connection().send_with_throttling(std::move(reply))
+    );
+    co_return;
+  }
+
+  DEBUGDPP("{}.{}: not completed, entering get_obc stage",
+	   *pg, *this, this_instance_id);
+  co_await ihref.enter_stage<interruptor>(client_pp(*pg).get_obc, *this);
+
+  DEBUGDPP("{}.{}: entered get_obc stage, about to wait_scrub",
+	   *pg, *this, this_instance_id);
+  if (int res = op_info.set_from_op(&*m, *pg->get_osdmap());
+      res != 0) {
+    co_await reply_op_error(pg, res);
+    co_return;
+  }
+  co_await ihref.enter_blocker(
+    *this, pg->scrubber, &decltype(pg->scrubber)::wait_scrub,
+    m->get_hobj());
+
+  DEBUGDPP("{}.{}: past scrub blocker, getting obc",
+	   *pg, *this, this_instance_id);
+  co_await pg->with_locked_obc(
+    m->get_hobj(), op_info,
+    [FNAME, this, pg, this_instance_id, &ihref] (
+      auto head, auto obc
+    ) -> interruptible_future<> {
+      DEBUGDPP("{}.{}: got obc {}, entering process stage",
+	       *pg, *this, this_instance_id, obc->obs);
+      return ihref.enter_stage<interruptor>(
+	client_pp(*pg).process, *this
+      ).then_interruptible(
+	[FNAME, this, pg, this_instance_id, obc, &ihref]() mutable {
+	  DEBUGDPP("{}.{}: in process stage, calling do_process",
```

Comment:
Can this be further simplified?
```
  DEBUGDPP("{}.{}: got obc {}, entering process stage",
           *pg, *this, this_instance_id, obc->obs);
  co_await ihref.enter_stage<interruptor>(client_pp(*pg).process, *this);

  DEBUGDPP("{}.{}: in process stage, calling do_process",
           *pg, *this, this_instance_id);
  co_await do_process(ihref, pg, obc, this_instance_id);
```

Reply:
Yes, but not now.  Converting that lambda to a coroutine has a catch that would need to be handled -- the lambda lifetime itself would have to outlive resulting future.

Reply:
Can you elaborate when we shouldn't conver a lambda to a coroutine right now? -- Do you mean the lifetime issue is related to the lambda passed into `pg->with_locked_obc()` ?

Reply:
@athanatos do you mean we need implement something like: seastar::coroutine::lambda